### DISCUSSION TITLE
Issues by stage5

### DISCUSSION THREAD

**Discussion by kolserdav** [atherdon/dash-back#72] [General]
Здесь напишу вопросы по [stage5.md](https://github.com/atherdon/dash-back/blob/main/docs/stages/stage5.md)

**Comment by kolserdav** [comment]
В файлах из `data` я не увидел практически никаких связей между схемами, кроме последней версии `expandable.json`, где articles можно заменить ключами на статьи. Также непонятно как должны быть связаны статьи с  тегами и ключевыми словами. Также `editors` должны иметь ссылку на пользователя или пользователь должен иметь ссылку (ключ - индекс) на едитора коим является. В общем нужно больше информации о целях бекенда. Например хорошо поможет утвержденный макет админки, чтобы ориентировать бек на будущий фронт.

**Reply by kolserdav** [reply]
Тз на фронт тоже прольет свет.

**Reply by atherdon** [reply]
данные в data - это был просто экспорт информации из excel файла, который марина сейчас использует как хранилище.
новой экспортной инфы сейчас нет. ее реально сделать, но нужно будет садится.

`Editors` пока не трогаем - потому что я не придумал еще как хранить информацию об authors и нужно ли это вообще делать в рамках нашей внутренней системы и бекенда.


у нас есть старая админка
![image](https://user-images.githubusercontent.com/1469198/109381910-cdc4c200-78e5-11eb-8a70-697f4b3ee044.png)

текущая версия
![image_2021-02-19_18-44-52](https://user-images.githubusercontent.com/1469198/109381975-3449e000-78e6-11eb-8def-47d65f90d9b8.png)
![image_2021-02-12_18-18-00](https://user-images.githubusercontent.com/1469198/109381985-4461bf80-78e6-11eb-9593-bc9d6e6a4a5f.png)

**Reply by atherdon** [reply]
еще у нас есть новые страницы для фронтенда, которые мы еще не перенесли в react:
https://github.com/atherdon/dashb/tree/main/frontend/dist

админка нужна чтобы получать доступ к статьям, которые отправляют авторы на выкладывание.
статьи бывают разных типов и категорий. под некоторые типы статей существуют отдельные редакторы.
кроме редаторов еще есть ассистент редактора - Марина, которая помогает сейчас с группировкой новых статей по типам(expandable3) и отправляет эти статьи другим редакторам ручками. наша система получается частично автоматизирует работу Марины и освободит ее немного от рутинной работы.

посути - вся логика внутри нашей админки сейчас крутится вокруг статей. просто эти статьи могут отличатся разными способами. и мы пытаемся сделать так - чтобы редакторы тратили меньше времени на выбор статьи и больше времени на выкладывание статей.

при это нам еще приходит много спама, поэтому нам нужно разграничение на `new submissions` `ordinary articles`, `evergreen` - потому что для их выкладывания требуются разные правила.

если обычную статью в которой топ автор себя промоутит - я могу выложить потому что он рок-звезда
то такую же статью от noname - я просто отклоню.
если человек продвигает бренд через бесплатный аккаунт - я тоже отклоню статью.
а если он создаст бренд аккаунт, и будет возможность что он скоро заплатит(при том что статья не будет треш) - ее можно выложить, если она платная.

поэтому разбивая по типам - мы упрощаем авторам decision-making. в этом главная цель.
статистика и т.д. - это скорее дополнительные плюшки, которые просто нужны для управления командой

**Reply by kolserdav** [reply]
как я понимаю, `dash-back` будет бекендом только для админки? если так то бек сайта будет добавлять в него ссылки на статьи для редактирования редакторами, а уже редактирование будет происходить с основным беком.

**Reply by atherdon** [reply]
Если кратко, то да. сейчас мы просто как второстепенный проект. Но зато тк у нас нету всего tech-debt, который есть у основного монолита - мы можем быстрее двигатся вперед и релизить новые фичи. потому что старой админке 19 месяцев - и она совсем не предназначена для той работы - которую мы делаем сейчас.

мы уже не просто выросли из штанишек - мы посути работаем в другой системе. если раньше была система: автор - статья - редактор.
то теперь есть несколько систем

- автор - статья - бесплатная - человек который форматирует - редактор.
- бренд - статья - форматирование - редактор - продвижение
- slack - автор - статья другого типа - много форматирования - редактор - выкладывание

**Comment by kolserdav** [comment]
Вот у нас статьи из примерных файлов
```
{
    'Draft url': 'https://app.hackernoon.com/drafts/4oJ3dI9QGwqwHuhBYT8F',
    '': '',
    v: 'v',
    email: 'mailto:xxx.com',
    Published: 'TRUE',
    Added: 4.09,
    Edited: 7.09,
    Published__1: 7.09,
    'Avg time story': 4,
    'Avg time ALL story': 6,
    __1: ''
  },
```
1. `url` - сылка на статью. Она постоянная незваисимо от изменения типа статьи?
2. `v` - что это?
3. `publised` - изменим на `boolean`
4. `added` - когда статья была добавлена. Сейчас нигде не прописан год и это не даёт возможности хранить как дату. Как дальше с этим жить?
5. `edited` - дата редактирования статьи. - // -
6. `published` - дата публикации. - // -
7. `avgTimeStory` - вроде дней от первого редактирования до публикации
8. `avfTimeAllStrory` - ароде дней от добавления статьи до публикации
9. `email` - служит для идентификации автора статьи?

По датам надо определиться, как `dash-back` будет получать данные?! Сейчас временно реализовано через `dash-upload-forms`. в будущем я так понимаю, будет правильным, если это будет заполняться программно, то есть при добавлении статьи на сайт основной бек будет делать запись о статье в базу `dash-back`. Значит мы уже сейчас можем рассчитывать, что дата `added` будет валидной и хранить её в `DateTime` а `edited`  и `published` записывать по умолчанию `null` и менять уже при редактировании и публиковании соответственно. __Тогда вопрос__, как быть с невалидными датами из `data`? Мы же не можем проставить 2020 там где месяц ещё не наступил и 2020 где наоборот - этот скрипт допустит много ошибок в соответствии даты фактической.

**Reply by atherdon** [reply]
- url - сылка на статью. Она постоянная незваисимо от изменения типа статьи?
ссылка у каждой статьи уникальная.

- v - что это?
это чекбокс, которым марина отмечала какие статьи уже готовы для удобства.

- publised - изменим на boolean
тоже похоже на чекбокс по функционалу, но со временем, мы сделаем это отдельной схемой.

- added - когда статья была добавлена. Сейчас нигде не прописан год и это не даёт возможности хранить как дату. Как дальше с этим жить?
пока можно оставить как есть. date у нас должны работать для статистики, но мне нужно время чтобы это нормально описывать.

- edited - дата редактирования статьи. - // -
тоже хочу оставить на следующие майлстоуны

- published - дата публикации. - // -
тоже хочу оставить на следующие майлстоуны

- avgTimeStory - вроде дней от первого редактирования до публикации
да, я хочу считать, сколько дней у нас занимает одна статья до полного выкладывания. сейчас мы хотим чтобы это число было дней 5. но сколько реально времени у нас это занимает - никто не знает

- avfTimeAllStrory - ароде дней от добавления статьи до публикации
тоже отдельный milestone

- email - служит для идентификации автора статьи?
я пока сомневаюсь что я хочу хранить emails в нашем бекенде. мы сохраняли emails тех, кому мы хотим написать и предложить нам заплатить. но хранение emails накладывает кучу проблем - i.e. в европе, если захотят - могут наложить штраф за то что emails были неправильно использованы.

раз мы зацепили тему дат и calculations - может для дат нам сделать отдельную схему - посути вводить их руками в будущем мы не будем. эта часть будет автоматическая

сс @kolserdav

**Reply by atherdon** [reply]
< Сейчас временно реализовано через dash-upload-forms. в будущем я так понимаю, будет правильным, если это будет заполняться программно, то есть при добавлении статьи на сайт основной бек будет делать запись о статье в базу dash-back.

все верно.

< Тогда вопрос, как быть с невалидными датами из data? Мы же не можем проставить 2020 там где месяц ещё не наступил и 2020 где наоборот - этот скрипт допустит много ошибок в соответствии даты фактической.

пока можно просто пропускать невалидные даты. или давайте сделаем отдельный таск, чтобы незабыть, и когда у нас накопится таких тасков 10 - мы их объединим в отдельный stage. 

сейчас мне лень думать об этом - мы решили 4 основные проблемы за последний месяц, но у меня теперь есть задача заставить редакторов использовать новый вариант - а люди обычно очень интертные.

если хотите - можем сделать так - как вам кажется будет лучше - я тут непротив.
но раздел работы с датами и статистикой - достаточно большой и я его даже не начал делать.
текущая версия находится в этом thread: https://github.com/atherdon/dash-back/discussions/60
но данных там сейчас очень мало.

**Comment by kolserdav** [comment]
>раз мы зацепили тему дат и calculations - может для дат нам сделать отдельную схему - посути вводить их руками в будущем мы не будем. эта часть будет автоматическая
>
да, можно убрать из статей даты и `avg...s`, вынести их в отдельную схему и привязать к каждой статье ссылку на её статистику.
>то есть при добавлении статьи на сайт основной бек будет делать запись о статье в базу dash-back. Значит мы уже сейчас можем рассчитывать, что дата added будет валидной и хранить её в DateTime а edited и published записывать по умолчанию null и менять уже при редактировании и публиковании соответственно.
>
С датами только так могу предложить. Однако, тогда даты из файлов не подойдут и их придётся записывать `null` и они там по моему все такие, а потом даже не знаю как их можно будет перенести, наверное только руками, но уже с указанием года.

>пока можно просто пропускать невалидные даты. или давайте сделаем отдельный таск, чтобы незабыть, и когда у нас накопится таких тасков 10 - мы их объединим в отдельный stage.
>
сейчас же нормализацию таблиц пытаемся делать, если даты не будут в нормальном формате, то в будущем нельзя будет добавить фильтры по дате добавления, редактирования и публикования.
>пока можно просто пропускать невалидные даты. или давайте сделаем отдельный таск, чтобы незабыть, и когда у нас накопится таких тасков 10 - мы их объединим в отдельный stage.
>
не думаю что к этому вопросу можно так легко относится. если с ним одним разбираясь пухнет голова, то если добавить к ней ещё 9 задач сверху - вряд ли это поможет.

если у вас сейчас нет времени заниматься этим вопросом, давайте тогда перенесём решение по датам скажем на понедельник?

Я вас услышал, но работаем не первый день, поэтому не хочется сделать "лишь бы закрыть таск", а так чтобы это потом не пришлось переделывать, всех денег не заработаешь, а нервы надо беречь)

**Reply by atherdon** [reply]
@kolserdav да, я могу в понедельник начать описание, связанное с датами и статистикой. 
Уже наше обсуждение продвинуло меня в мыслях вперед. 

Я просто помню что у меня на листиках где то была информация по статистике. но я это писал в июле еще - за это время много всего изменилось.

с `date` согласен - наверное все проблемы идут от того, что данные то были экспортированы в `csv` из экселя - наверное он нам все и портит.

может нам просто переделать значения `date` в папке `\data` таким образом, чтобы нам было удобнее?
потому что при добавлении новых записей - мы даты будем сами позже записывать, в удобном для нас формате.

я согласен - что архитектуру на абы-как лучше не делать. оно потом обязательно вылезет в самый неподходящий момент. плюс еще в том - у нас сейчас есть время подумать и сделать хорошо.

**Comment by kolserdav** [comment]
>может нам просто переделать значения date в папке \data таким образом, чтобы нам было удобнее?
потому что при добавлении новых записей - мы даты будем сами позже записывать, в удобном для нас формате.
>
Да, а вот как их переделать?! Время, к примеру, можно везде любое поставить, что тоже не совсем гуд, но вот когда год не указан, программа не сможет его придумать.

**Reply by kolserdav** [reply]
Может есть какой нибудь апи основного бекенда откуда можно трансфер сделать?

**Reply by atherdon** [reply]
< Да, а вот как их переделать?!

Я имелввиду переделать до такого формата - который будет амудобен.
если нам нужен год - давайте его вставим, например.

< Может есть какой нибудь апи основного бекенда откуда можно трансфер сделать?

апи убекенда есть, я планирую связать наши проекты со временем, но не сейчас. 
думаю что мне проще будет попросить девов сделать кнопку экспорта данных в json формате. 

вначале мне нужно заставить всх редакторов начать использовать новую админку. это 6 человек + Марина(но она тосогласится - потому что мы ее работу сильно упрощаем)

а потом я поймаю врем, когда у наших программистов есть свободное время и попрощу их помощи

**Reply by atherdon** [reply]
@kolserdav теперь я понял в чем наша проблема с date.
https://github.com/atherdon/dash-back/blob/main/data/brands.js#L23

я просто должен ручками заменить эти значения на дату в нормальном формате.
можете скинуть пример в каком формате нам лучше хранить ее? а я данные заменю - не прблема с моей стороны

плюс я нашел прикольный модуль, может он нам тоже подойдет: https://jacobruiz.com/blog/2019/6/11/why-use-custom-date-scalar-types-in-graphql

**Reply by kolserdav** [reply]
В иделе лучше в таком "2019-06-12T03:29:56.901Z"

**Reply by atherdon** [reply]
принял. тогда считаем что у нас данные в этом формате и приходят. данные в папке data - я переделаю.

если что, мы можем потом заюзать модуль momentjs - если нам нужно будет это для работы с датами.

**Reply by kolserdav** [reply]
Хорошо, а сущность статья в одну схему делаем? Чтобы от клонов избавиться.

**Reply by atherdon** [reply]
конечно, это и был изначальный план :) даты эти нас просто решили поотвлекать

**Reply by kolserdav** [reply]
И еще вопрос, какие фильтры и сортировки , может группировки, на множественный get нужны?

**Reply by kolserdav** [reply]
Пагинацию не спрашиваю, так как естественно да)

**Reply by atherdon** [reply]
сейчас мне кажется группировка сортировка будет по типам статей и по published=false.

тоесть мы же будем выводить статьи типа brand - на отдельной странице
статьи типа topAuthors - на отдельной странице
и т.д.

наверное это первое что приходит мне на ум.
я могу создать отдельный таск про фильтры и сортировки и подумать об этом чуть позже. сразу мне сложно ответить

**Reply by atherdon** [reply]
Я по быстрому обновил дату в https://github.com/atherdon/dash-back/blob/main/data/brands.js
могу чуть позже сделать тоже самое и для других файлов.

такой вопрос - давайте я создам таск, и внутри него мы обсудим как я могу "полечить" файлы внутри `/data` чтобы вам было удобнее работать. я об этой проблеме как то раньше не думал. но теперь есть время подумать и сделать все так, как вам будет удобнее

**Reply by kolserdav** [reply]
И ещё делаем ли отсечение лишних селектов к базе, то есть чтобы бек просил у базы только те поля которые просит фронт? Это только уменьшит потребление RAM.

**Reply by kolserdav** [reply]
В общем, то что обговорили и sync скрипта наполнения разных типов статей с новыми датами и sync dash-upload-forms будет 2 дня.

**Reply by atherdon** [reply]
создал проект на FH.
вопрос выше вынесу в отдельный таск

**Comment by kolserdav** [comment]
>плюс я нашел прикольный модуль, может он нам тоже подойдет: https://jacobruiz.com/blog/2019/6/11/why-use-custom-date-scalar-types-in-graphql
>
Он нам не нужен. По сути он делает тоже самое что у нас делает `date.toISOString()`.

**Reply by kolserdav** [reply]
Был неправ, он очень хорошо пойдет в версии 5.1.2 @atherdon  спасибо за ссылку)

**Reply by kolserdav** [reply]
На выходные данные у нас он особо не влияет, однако входные данные проверяет очень даже неплохо, но вот проблема `react-apollo-form` не воспринимает формат `DateTime` и просто пропускает эти поля. Получается в `dash-upload-forms` нельзя его использовать, что делать?

**Reply by kolserdav** [reply]
Придется 'postOneArticle' и 'updateOneArticle' в `dash-upload-form` переделать на хуках, завтра только готово будет.

**Comment by kolserdav** [comment]
@atherdon у Вас так быстро получилось даты в `brands.json` проставить, может и в `top-authors.json` сделаете, чтобы сразу двух типов `articles` добавить?

**Reply by atherdon** [reply]
да, я хотел это сделать вчера но у меня интернет отпал

**Reply by atherdon** [reply]
я пока просто добавил копии, чтобы не крашился json. чуть позже сделаю информацию реальнее - чтобы мы могли тестить это проще

**Comment by kolserdav** [comment]
И еще по `email` в `articles` я не понял, убираем или пока оставляем?

**Reply by atherdon** [reply]
пока убираем

**Reply by kolserdav** [reply]
ок. там наверное потом логичней будет `id` пользователя, данные которого будут где-то в защищенном месте.

**Comment by atherdon** [comment]
как минимум я доволен тем фактом, что формы мы вынесли в отдельный проект.
это сразу нам упростило много всего.
представьте если бы сейчас эта проблема появилась на основном фронте

>
